README.md
# gem-compiler

A RubyGems plugin that generates binary (pre-compiled) gems.

[![Gem Version](https://img.shields.io/gem/v/gem-compiler.svg)](https://rubygems.org/gems/gem-compiler)
[![Travis status](https://img.shields.io/travis/luislavena/gem-compiler/master.svg)](https://travis-ci.org/luislavena/gem-compiler)
[![AppVeyor status](https://img.shields.io/appveyor/ci/luislavena/gem-compiler/master.svg)](https://ci.appveyor.com/project/luislavena/gem-compiler)
[![Code Climate](https://img.shields.io/codeclimate/github/luislavena/gem-compiler.svg)](https://codeclimate.com/github/luislavena/gem-compiler)

- [home](https://github.com/luislavena/gem-compiler)
- [bugs](https://github.com/luislavena/gem-compiler/issues)

## Description

`gem-compiler` is a RubyGems plugin that helps generates binary gems from
already existing ones without altering the original source code. It compiles
Ruby C extensions and bundles the result into a new gem.

It uses an *outside-in* approach and leverages on existing RubyGems code to
do it.

The result of the compilation is a binary gem built for your current platform,
skipping the need of a compiler toolchain when installing it.

## Installation

To install gem-compiler you need to use RubyGems:

    $ gem install gem-compiler

Which will fetch and install the plugin. After that the `compile` command
will be available through `gem`.

## Usage

As requirement, gem-compiler can only compile local gems, either one you have
generated from your projects or previously downloaded.

### Fetching a gem

If you don't have the gem locally, you can use `fetch` to retrieve it first:

    $ gem fetch yajl-ruby --platform=ruby
    Fetching: yajl-ruby-1.1.0.gem (100%)
    Downloaded yajl-ruby-1.1.0

Please note that I was explicit about which platform to fetch. This will
avoid RubyGems attempt to download any existing binary gem for my current
platform.

### Compiling a gem

You need to tell RubyGems the filename of the gem you want to compile:

    $ gem compile yajl-ruby-1.1.0.gem

The above command will unpack, compile any existing extensions found and
repackage everything as a binary gem:

    Unpacking gem: 'yajl-ruby-1.1.0' in temporary directory...
    Building native extensions.  This could take a while...
      Successfully built RubyGem
      Name: yajl-ruby
      Version: 1.1.0
      File: yajl-ruby-1.1.0-x86-mingw32.gem

This new gem do not require a compiler, as shown when locally installed:

    C:\> gem install --local yajl-ruby-1.1.0-x86-mingw32.gem
    Successfully installed yajl-ruby-1.1.0-x86-mingw32
    1 gem installed

There are native gems that will invalidate their own specification after
compile process completes. This will not permit them be repackaged as binary
gems. To workaround this problem you have the option to *prune* the package
process:

    $ gem fetch nokogiri --platform=ruby
    Fetching: nokogiri-1.6.6.2.gem (100%)
    Downloaded nokogiri-1.6.6.2

    $ gem compile nokogiri-1.6.6.2.gem --prune
    Unpacking gem: 'nokogiri-1.6.6.2' in temporary directory...
    Building native extensions.  This could take a while...
      Successfully built RubyGem
      Name: nokogiri
      Version: 1.6.6.2
      File: nokogiri-1.6.6.2-x86_64-darwin-12.gem

    $ gem install --local nokogiri-1.6.6.2-x86_64-darwin-12.gem
    Successfully installed nokogiri-1.6.6.2-x86_64-darwin-12
    1 gem installed

#### Restrictions of binaries

Gems compiled with `gem-compiler` will be lock to the version of Ruby used
to compile them.

This means that a gem compiled under Ruby 2.2 could only be installed under
Ruby 2.2.

You can disable this by using `--no-abi-lock` option during compilation:

    $ gem compile yajl-ruby-1.1.0.gem --no-abi-lock

**Warning**: this is not recommended since different versions of Ruby might
expose different options on the API. The binary might be expecting specific
features not present in the version of Ruby you're installing the binary gem
into.

### Compiling from Rake

Most of the times, as gem developer, you would like to generate both kind of
gems at once. For that purpose, you can add a task for Rake similar to the
one below:

```ruby
desc "Generate a pre-compiled native gem"
task "gem:native" => ["gem"] do
  sh "gem compile #{gem_file}"
end
```

Of course, that assumes you have a task `gem` that generates the base gem
required.

## Requirements

### Ruby and RubyGems

It's assumed you have Ruby and RubyGems installed. gem-compiler requires
RubyGems 1.8.x to properly work.

If you don't have RubyGems 1.8.x, you can upgrade by running:

    $ gem update --system

### A compiler

In order to compile a gem, you need a compiler toolchain installed. Depending
on your Operating System you will have one already installed or will require
additional steps to do it. Check your OS documentation about getting the
right one.

### If you're using Windows

For those using RubyInstaller-based builds, you will need to download the
DevKit from their [downloads page](http://rubyinstaller.org/downloads)
and follow the [installation instructions](https://github.com/oneclick/rubyinstaller/wiki/Development-Kit).

To be sure your installation of Ruby is based on RubyInstaller, execute at
the command prompt:

    C:\> ruby --version

And from the output:

    tcs-ruby 1.9.3p196 (2012-04-21, TCS patched 2012-04-21) [i386-mingw32]

If you see `mingw32`, that means you're using a RubyInstaller build
(MinGW based).

Once you have installed DevKit into the version of ruby that you wish to build against,
the "gem compile" command might fail with the error message,

    ERROR:  While executing gem ... (Gem::Installer::ExtensionBuildError)
    ERROR: Failed to build gem native extensions.
    
The solution is to ensure your path is enhanced with the DevKit tools.
Execute the following from your DevKit home directory:

    devkitvars.bat
    
Re-run "gem compile gemname-0.0.0.gem --prune" and it should be successful.

## Differences with rake-compiler

[rake-compiler](https://github.com/luislavena/rake-compiler) has provided to
Ruby library authors a *tool* for compiling extensions and generating binary
gems of their libraries.

You can consider rake-compiler's approach be an *inside-out* process. To do
its magic, it requires library authors to modify their source code, adjust
some structure and learn a series of commands.

While the ideal scenario is using a tool like rake-compiler that endorses
*convention over configuration*, is not humanly possible change all the
projects by snapping your fingers :wink:

## License

[The MIT License](LICENSE)


History.md
# gem-compiler

## Unreleased

## 0.6.0 (2017-06-25)

- Drop support for any Ruby version prior to 2.1.0
- Solve RubyGems 2.5 deprecation warnings
- Use Travis to automate new releases
- CI: Update test matrix (Travis and AppVeyor)

## 0.5.0 (2016-04-24)

- Drop support for any Ruby version prior to 2.0.0
- Workaround shortname directories on Windows. Thanks to @mbland (#17 & #19)
- Validate both Ruby and RubyGems versions defined in gemspec
- Ensure any RubyGems' `pre_install` hooks are run at extension compilation (#18)
- Lock compile gems to Ruby's ABI version which can be disabled using
  `--no-abi-lock` option (#11)

## 0.4.0 (2015-07-18)

- Introduce `--prune` option to cleanup gemspecs. Thanks to @androbtech [#13]
- Test builds on both Travis (Linux) and AppVeyor (Windows) environments.

## 0.3.0 (2014-04-19)

- Support RubyGems 2.2.x thanks to @drbrain
- Minor reorganization to make testing on Travis more easy

## 0.2.0 (2013-04-28)

- Support RubyGems 2.0.0 thanks to @mgoggin [#6]

## 0.1.1 (2012-05-07)

- Loose requirements to allow installation on Ruby 1.8.7 or greater. You
  still need RubyGems 1.8.24

## 0.1.0 (2012-05-06)

- Initial public release, extracted from internal project.


test/test_helper.rb
require "rubygems/test_case"

class Gem::TestCase
  ##
  # Reset RubyGems platform to original one. Useful when testing platform
  # specific features (like compiled extensions)

  def util_reset_arch
    util_set_arch @orig_arch
  end

  ##
  # Create a real gem and return the path to it.

  def util_bake_gem(name = "a", *extra, &block)
    files = ["lib/#{name}.rb"].concat(extra)

    spec = new_spec name, "1", nil, files, &block

    File.join @tempdir, "gems", "#{spec.full_name}.gem"
  end

  ##
  # Add a fake extension to provided spec and accept an optional script.
  # Default to no-op if none is provided.

  def util_fake_extension(spec, name = "a", script = nil)
    mkrf_conf = File.join("ext", name, "mkrf_conf.rb")

    spec.extensions << mkrf_conf

    dir = spec.gem_dir
    FileUtils.mkdir_p dir

    Dir.chdir dir do
      FileUtils.mkdir_p File.dirname(mkrf_conf)
      File.open mkrf_conf, "w" do |f|
        if script
          f.write script
        else
          f.write <<-EOF
            File.open 'Rakefile', 'w' do |rf| rf.puts "task :default" end
          EOF
        end
      end
    end
  end

  ##
  # Constructor of custom configure script to be used with
  # +util_fake_extension+
  #
  # Provided +target+ will be used to fake an empty file at default task

  def util_custom_configure(target)
    <<-EO_MKRF
      File.open("Rakefile", "w") do |f|
        f.puts <<-EOF
          task :default do
            lib_dir = ENV["RUBYARCHDIR"] || ENV["RUBYLIBDIR"]
            touch File.join(lib_dir, #{target.inspect})
          end
        EOF
      end
    EO_MKRF
  end

  ##
  # Return the metadata (spec) from the supplied filename. IO from filename
  # is closed automatically

  def util_read_spec(filename)
    unless Gem::VERSION >= "2.0.0"
      io = File.open(filename, "rb")
      Gem::Package.open(io, "r") { |x| x.metadata }
    else
      Gem::Package.new(filename).spec
    end
  end
end


test/rubygems/test_gem_compiler.rb
require "rubygems/test_case"
require "rubygems/compiler"

class TestGemCompiler < Gem::TestCase
  def setup
    super

    # unset GEM_PATH so `rake` is found during compilation of extensions
    ENV.delete("GEM_PATH")

    @output_dir = File.join @tempdir, 'output'
    FileUtils.mkdir_p @output_dir
  end

  def test_compile_no_extensions
    gem_file = util_bake_gem

    compiler = Gem::Compiler.new(gem_file, :output => @output_dir)

    e = assert_raises Gem::Compiler::CompilerError do
      use_ui @ui do
        compiler.compile
      end
    end

    assert_equal "There are no extensions to build on this gem file.",
                 e.message
  end

  def test_compile_non_ruby
    gem_file = util_bake_gem { |s| s.platform = Gem::Platform::CURRENT }

    compiler = Gem::Compiler.new(gem_file, :output => @output_dir)

    e = assert_raises Gem::Compiler::CompilerError do
      use_ui @ui do
        compiler.compile
      end
    end

    assert_equal "The gem file seems to be compiled already.", e.message
  end

  def test_compile_pre_install_hooks
    util_reset_arch

    artifact = "foo.#{RbConfig::CONFIG["DLEXT"]}"

    gem_file = util_bake_gem("foo") { |s|
      util_fake_extension s, "foo", util_custom_configure(artifact)
    }

    hook_run = false

    Gem.pre_install do |installer|
      hook_run = true
      true
    end

    compiler = Gem::Compiler.new(gem_file, :output => @output_dir)

    use_ui @ui do
      compiler.compile
    end

    assert hook_run, "pre_install hook not run"
  end

  def test_compile_required_ruby
    gem_file = util_bake_gem("old_required") { |s| s.required_ruby_version = "= 1.4.6" }

    compiler = Gem::Compiler.new(gem_file, :output => @output_dir)

    e = assert_raises Gem::InstallError do
      use_ui @ui do
        compiler.compile
      end
    end

    assert_equal "old_required requires Ruby version = 1.4.6.", e.message
  end

  def test_compile_required_rubygems
    gem_file = util_bake_gem("old_rubygems") { |s| s.required_rubygems_version = "< 0" }

    compiler = Gem::Compiler.new(gem_file, :output => @output_dir)

    e = assert_raises Gem::InstallError do
      use_ui @ui do
        compiler.compile
      end
    end

    assert_equal "old_rubygems requires RubyGems version < 0. " +
                   "Try 'gem update --system' to update RubyGems itself.", e.message
  end

  def test_compile_succeed
    util_set_arch "i386-mingw32"

    gem_file = util_bake_gem { |spec|
      util_fake_extension spec
    }

    compiler = Gem::Compiler.new(gem_file, :output => @output_dir)

    use_ui @ui do
      compiler.compile
    end

    out = @ui.output.split "\n"

    assert_match %r|Unpacking gem: 'a-1' in temporary directory...|,
                 out.shift

    assert_path_exists File.join(@output_dir, "a-1-x86-mingw32.gem")
  end

  def test_compile_succeed_using_prune
    name = 'a'

    artifact = "#{name}.#{RbConfig::CONFIG["DLEXT"]}"
    old_spec = ''

    gem_file = util_bake_gem(name, 'ports/to_be_deleted_during_ext_build.patch') { |spec|
      old_spec = spec
      util_fake_extension spec, name, <<-EOF
        require 'fileutils'
        FileUtils.rm File.expand_path(File.join(File.dirname(__FILE__), '../../ports/to_be_deleted_during_ext_build.patch'))
        #{util_custom_configure(artifact)}
      EOF
    }

    compiler = Gem::Compiler.new(gem_file, :output => @output_dir, :prune => true)
    output_gem = nil

    use_ui @ui do
      output_gem = compiler.compile
    end

    assert_path_exists File.join(@output_dir, output_gem)
    actual_spec = util_read_spec File.join(@output_dir, output_gem)

    refute actual_spec.files.include? "ports/to_be_deleted_during_ext_build.patch"
  end

  def test_compile_bundle_artifacts
    util_reset_arch

    artifact = "foo.#{RbConfig::CONFIG["DLEXT"]}"

    gem_file = util_bake_gem("foo") { |s|
      util_fake_extension s, "foo", util_custom_configure(artifact)
    }

    compiler = Gem::Compiler.new(gem_file, :output => @output_dir)
    output_gem = nil

    use_ui @ui do
      output_gem = compiler.compile
    end

    assert_path_exists File.join(@output_dir, output_gem)
    spec = util_read_spec File.join(@output_dir, output_gem)

    assert_includes spec.files, "lib/#{artifact}"
  end

  # We need to check that tempdir paths that contain spaces as are handled
  # properly on Windows. In some cases, Dir.tmpdir may returned shortened
  # versions of these components, e.g.  "C:/Users/JOHNDO~1/AppData/Local/Temp"
  # for "C:/Users/John Doe/AppData/Local/Temp".
  def test_compile_bundle_artifacts_path_with_spaces
    skip("only necessary to test on Windows") unless Gem.win_platform?
    old_tempdir = @tempdir
    old_output_dir = @output_dir

    old_tmp = ENV["TMP"]
    old_temp = ENV["TEMP"]
    old_tmpdir = ENV["TMPDIR"]

    # We want to make sure Dir.tmpdir returns the path containing "DIRWIT~1"
    # so that we're testing whether the compiler expands the path properly. To
    # do this, "dir with spaces" must not be the last path component.
    #
    # This is because Dir.tmpdir calls File.expand_path on ENV[TMPDIR] (or
    # ENV[TEMP], etc.). When "DIRWIT~1" is the last component,
    # File.expand_path will expand this to "dir with spaces". When it's not
    # the last component, it will leave "DIRWIT~1" as-is.
    @tempdir = File.join(@tempdir, "dir with spaces", "tmp")
    FileUtils.mkdir_p(@tempdir)
    @tempdir = File.join(old_tempdir, "DIRWIT~1", "tmp")

    @output_dir = File.join(@tempdir, "output")
    FileUtils.mkdir_p(@output_dir)

    ["TMP", "TEMP", "TMPDIR"].each { |varname| ENV[varname] = @tempdir }

    util_reset_arch

    artifact = "foo.#{RbConfig::CONFIG["DLEXT"]}"

    gem_file = util_bake_gem("foo") { |s|
      util_fake_extension s, "foo", util_custom_configure(artifact)
    }

    compiler = Gem::Compiler.new(gem_file, :output => @output_dir)
    output_gem = nil

    use_ui @ui do
      output_gem = compiler.compile
    end

    assert_path_exists File.join(@output_dir, output_gem)
    spec = util_read_spec File.join(@output_dir, output_gem)

    assert_includes spec.files, "lib/#{artifact}"
  ensure
    if Gem.win_platform?
      FileUtils.rm_rf @tempdir

      ENV["TMP"] = old_tmp
      ENV["TEMP"] = old_temp
      ENV["TMPDIR"] = old_tmpdir

      @tempdir = old_tempdir
      @output_dir = old_output_dir
    end
  end

  def test_compile_lock_ruby_abi
    util_reset_arch

    ruby_abi = RbConfig::CONFIG["ruby_version"]
    artifact = "foo.#{RbConfig::CONFIG["DLEXT"]}"

    gem_file = util_bake_gem("foo") { |s|
      util_fake_extension s, "foo", util_custom_configure(artifact)
    }

    compiler = Gem::Compiler.new(gem_file, :output => @output_dir)
    output_gem = nil

    use_ui @ui do
      output_gem = compiler.compile
    end

    spec = util_read_spec File.join(@output_dir, output_gem)

    assert_equal spec.required_ruby_version, Gem::Requirement.new("~> #{ruby_abi}")
  end

  def test_compile_no_lock_ruby_abi
    util_reset_arch

    artifact = "foo.#{RbConfig::CONFIG["DLEXT"]}"

    gem_file = util_bake_gem("foo") { |s|
      util_fake_extension s, "foo", util_custom_configure(artifact)
    }

    compiler = Gem::Compiler.new(gem_file, :output => @output_dir, :no_abi_lock => true)
    output_gem = nil

    use_ui @ui do
      output_gem = compiler.compile
    end

    spec = util_read_spec File.join(@output_dir, output_gem)

    assert_equal spec.required_ruby_version, Gem::Requirement.new(">= 0")
  end

  ##
  # Reset RubyGems platform to original one. Useful when testing platform
  # specific features (like compiled extensions)

  def util_reset_arch
    util_set_arch @orig_arch
  end

  ##
  # Create a real gem and return the path to it.

  def util_bake_gem(name = "a", *extra, &block)
    files = ["lib/#{name}.rb"].concat(extra)

    spec = new_spec name, "1", nil, files, &block

    File.join @tempdir, "gems", "#{spec.full_name}.gem"
  end

  ##
  # Add a fake extension to provided spec and accept an optional script.
  # Default to no-op if none is provided.

  def util_fake_extension(spec, name = "a", script = nil)
    mkrf_conf = File.join("ext", name, "mkrf_conf.rb")

    spec.extensions << mkrf_conf

    dir = spec.gem_dir
    FileUtils.mkdir_p dir

    Dir.chdir dir do
      FileUtils.mkdir_p File.dirname(mkrf_conf)
      File.open mkrf_conf, "w" do |f|
        if script
          f.write script
        else
          f.write <<-EOF
            File.open 'Rakefile', 'w' do |rf| rf.puts "task :default" end
          EOF
        end
      end
    end
  end

  ##
  # Constructor of custom configure script to be used with
  # +util_fake_extension+
  #
  # Provided +target+ will be used to fake an empty file at default task

  def util_custom_configure(target)
    <<-EO_MKRF
      File.open("Rakefile", "w") do |f|
        f.puts <<-EOF
          task :default do
            lib_dir = ENV["RUBYARCHDIR"] || ENV["RUBYLIBDIR"]
            touch File.join(lib_dir, #{target.inspect})
          end
        EOF
      end
    EO_MKRF
  end

  ##
  # Return the metadata (spec) from the supplied filename. IO from filename
  # is closed automatically

  def util_read_spec(filename)
    unless Gem::VERSION >= "2.0.0"
      io = File.open(filename, "rb")
      Gem::Package.open(io, "r") { |x| x.metadata }
    else
      Gem::Package.new(filename).spec
    end
  end
end


test/rubygems/test_gem_commands_compile_command.rb
require "rubygems/test_case"
require "rubygems/commands/compile_command"
require "rubygems/package"

class TestGemCommandsCompileCommand < Gem::TestCase
  def setup
    super

    @cmd = Gem::Commands::CompileCommand.new
  end

  def test_execute_no_gem
    @cmd.options[:args] = []

    e = assert_raises Gem::CommandLineError do
      use_ui @ui do
        @cmd.execute
      end
    end

    assert_match %r{Please specify a gem file on the command line}, e.message
  end
end


lib/rubygems/commands/compile_command.rb
require "rubygems/command"

class Gem::Commands::CompileCommand < Gem::Command
  def initialize
    super "compile", "Create binary pre-compiled gem",
          output: Dir.pwd

    add_option "--prune", "Clean non-existing files during re-packaging" do |value, options|
      options[:prune] = true
    end

    add_option "-N", "--no-abi-lock", "Do not lock compiled Gem to Ruby's ABI" do |value, options|
      options[:no_abi_lock] = true
    end
  end

  def arguments
    "GEMFILE       path to the gem file to compile"
  end

  def usage
    "#{program_name} GEMFILE"
  end

  def execute
    gemfile = options[:args].shift

    # no gem, no binary
    unless gemfile
      raise Gem::CommandLineError,
            "Please specify a gem file on the command line (e.g. #{program_name} foo-0.1.0.gem)"
    end

    require "rubygems/compiler"

    compiler = Gem::Compiler.new(gemfile, options)
    compiler.compile
  end
end


lib/rubygems/compiler.rb
require "fileutils"
require "rbconfig"
require "tmpdir"
require "rubygems/installer"

if Gem::VERSION >= "2.0.0"
  require "rubygems/package"
else
  require "rubygems/builder"
end

class Gem::Compiler
  include Gem::UserInteraction

  # raise when there is a error
  class CompilerError < Gem::InstallError; end

  attr_reader :target_dir, :options

  def initialize(gemfile, _options = {})
    @gemfile = gemfile
    @output_dir = _options.delete(:output)
    @options = _options
  end

  def compile
    unpack

    build_extensions

    artifacts = collect_artifacts

    # build a new gemspec from the original one
    gemspec = installer.spec.dup

    adjust_gemspec_files gemspec, artifacts

    # generate new gem and return new path to it
    repackage gemspec
  ensure
    cleanup
  end

  private

  def adjust_gemspec_files(gemspec, artifacts)
    # remove any non-existing files
    if @options[:prune]
      gemspec.files.reject! { |f| !File.exist?("#{target_dir}/#{f}") }
    end

    # add discovered artifacts
    artifacts.each do |path|
      # path needs to be relative to target_dir
      file = path.sub("#{target_dir}/", "")

      debug "Adding '#{file}' to gemspec"
      gemspec.files.push file
    end
  end

  def build_extensions
    # run pre_install hooks
    if installer.respond_to?(:run_pre_install_hooks)
      installer.run_pre_install_hooks
    end

    installer.build_extensions
  end

  def cleanup
    FileUtils.rm_rf tmp_dir
  end

  def collect_artifacts
    # determine build artifacts from require_paths
    dlext = RbConfig::CONFIG["DLEXT"]
    lib_dirs = installer.spec.require_paths.join(",")

    Dir.glob("#{target_dir}/{#{lib_dirs}}/**/*.#{dlext}")
  end

  def info(msg)
    say msg if Gem.configuration.verbose
  end

  def debug(msg)
    say msg if Gem.configuration.really_verbose
  end

  def installer
    @installer ||= prepare_installer
  end

  def prepare_installer
    # RubyGems 2.5 specifics
    unpack_options = options.dup.merge(unpack: true)
    if Gem::Installer.respond_to?(:at)
      installer = Gem::Installer.at(@gemfile, unpack_options)
    else
      installer = Gem::Installer.new(@gemfile, options.dup.merge(unpack: true))
    end

    # RubyGems 2.2 specifics
    if installer.spec.respond_to?(:full_gem_path=)
      installer.spec.full_gem_path = @target_dir
      installer.spec.extension_dir = File.join(@target_dir, "lib")
    end

    # Ensure Ruby version is met
    if installer.respond_to?(:ensure_required_ruby_version_met)
      installer.ensure_required_ruby_version_met
    end

    # Check version of RubyGems (just in case)
    if installer.respond_to?(:ensure_required_rubygems_version_met)
      installer.ensure_required_rubygems_version_met
    end

    # Hmm, gem already compiled?
    if installer.spec.platform != Gem::Platform::RUBY
      raise CompilerError,
            "The gem file seems to be compiled already."
    end

    # Hmm, no extensions?
    if installer.spec.extensions.empty?
      raise CompilerError,
            "There are no extensions to build on this gem file."
    end

    installer
  end

  def repackage(gemspec)
    # clear out extensions from gemspec
    gemspec.extensions.clear

    # adjust platform
    gemspec.platform = Gem::Platform::CURRENT

    # adjust version of Ruby
    unless @options[:no_abi_lock]
      ruby_abi = RbConfig::CONFIG["ruby_version"]
      gemspec.required_ruby_version = "~> #{ruby_abi}"
    end

    # build new gem
    output_gem = nil

    Dir.chdir target_dir do
      output_gem = if defined?(Gem::Builder)
                     Gem::Builder.new(gemspec).build
                   else
                     Gem::Package.build(gemspec)
                   end
    end

    unless output_gem
      raise CompilerError,
            "There was a problem building the gem."
    end

    # move the built gem to the original output directory
    FileUtils.mv File.join(target_dir, output_gem), @output_dir

    # return the path of the gem
    output_gem
  end

  def tmp_dir
    @tmp_dir ||= Dir.glob(Dir.mktmpdir).first
  end

  def unpack
    basename = File.basename(@gemfile, ".gem")
    @target_dir = File.join(tmp_dir, basename)

    # unpack gem sources into target_dir
    # We need the basename to keep the unpack happy
    info "Unpacking gem: '#{basename}' in temporary directory..."
    installer.unpack(@target_dir)
  end
end
